constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,69 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+// AdAAddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// Both an object and an array can have values not described in the schema.
+	// A blank Type could be anything, including an object or array, so we should
+	// treat it the same as those types.
+	if sch.Type == "" || sch.Type == "object" || sch.Type == "array" {
+		trueBool := true
+		sch.XPreserveUnknownFields = &trueBool
+	}
+
+	if sch.Properties != nil {
+		for k, v := range sch.Properties {
+			if err := AddPreserveUnknownFields(&v); err != nil {
+				return err
+			}
+
+			// As v is not a pointer, we need to set the contents of v back into the original data structure
+			sch.Properties[k] = v
+		}
+	}
+
+	if sch.Items != nil {
+		if sch.Items.Schema != nil {
```

Comment:
Do we know the rules for when Items.Schema vs Items.JSONSchemas are set?

We should require at least one of them to be set in order to satisfy structural schema. It looks like the spec requires `items` to be set, though maybe there's a subtlety I'm missing:

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema

Reply:
JSONSchema allows one to set multiple different schemas under `items` as a way of providing per-field [tuple validation](https://json-schema.org/understanding-json-schema/reference/array.html#tuple-validation).  

I'm not seeing anything in the k8s spec that overtly disallows this, but I did some digging with SourceGraph and I found that the api server runs this [validate()](https://github.com/kubernetes/apiextensions-apiserver/blob/7537a765248141dadcfd216ca03c0a09260c3ae4/pkg/apis/apiextensions/validation/validation.go#L997) function at each as it [recurses through a CRD OpenAPI Schema](https://github.com/kubernetes/apiextensions-apiserver/blob/7537a765248141dadcfd216ca03c0a09260c3ae4/pkg/apis/apiextensions/validation/validation.go#L740).  This logic [disallows JSONSchemas](https://github.com/kubernetes/apiextensions-apiserver/blob/7537a765248141dadcfd216ca03c0a09260c3ae4/pkg/apis/apiextensions/validation/validation.go#L1049), requiring that its length be zero.

```golang
if schema.Items != nil && len(schema.Items.JSONSchemas) != 0 {
	allErrs = append(allErrs, field.Forbidden(fldPath.Child("items"), "items must be a schema object and not an array"))
}
```

Reply:
Max and I discussed and have decided that, since JSONSchemas is not supported by the API server (and wasn't as far back as k8s 1.14), we shouldn't take on supporting it in our code.  Instead, we'll return an error if we find JSONSchemas to be non `nil`.